### ISSUE TITLE
Query reusing graph clause variable in statement pattern returns incorrect result

### ISSUE THREAD

**Issue by nguyenm100** [eclipse/rdf4j#4659]
### Current Behavior

running the following query for a particular dataset produces different results on rdf4j (4.2.3) vs graphdb (10.2.2 which uses rdf4j 4.2.3)

query:
```
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix owl: <http://www.w3.org/2002/07/owl#>

select distinct ?modeliri ?modelLabel {
    graph ?modeliri {
        ?modeliri a owl:Ontology ;
        rdfs:label ?modelLabel .
    }
}
```

data:
```
<http://something/s> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2002/07/owl#Ontology> <http://something/s> .
<http://something/s> <http://www.w3.org/2000/01/rdf-schema#label> "s-label-something"@en <http://something/s> .
<http://something/s> <http://www.w3.org/2000/01/rdf-schema#label> "s-label-somethingelse"@en <http://somethingelse/c> .
```

### Expected Behavior

graphdb appears to be correct.

### Steps To Reproduce

1. rdf4j running on wsl docker, graphdb desktop on windows
2. insert quads:
```
<http://something/s> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2002/07/owl#Ontology> <http://something/s> .
<http://something/s> <http://www.w3.org/2000/01/rdf-schema#label> "s-label-something"@en <http://something/s> .
<http://something/s> <http://www.w3.org/2000/01/rdf-schema#label> "s-label-somethingelse"@en <http://somethingelse/c> .
```
3. run query:
```
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix owl: <http://www.w3.org/2002/07/owl#>

select distinct ?modeliri ?modelLabel {
    graph ?modeliri {
        ?modeliri a owl:Ontology ;
        rdfs:label ?modelLabel .
    }
}
```
4. expect 1 row on rdf4j 4.2.3 but received 2.
```
<http://something/s> | "s-label-something"@en
<http://something/s> | "s-label-somethingelse"@en
```
graphdb returns 1 (as expected)
```
<http://something/s> | "s-label-something"@en


```

### Version

4.2.3

### Are you interested in contributing a solution yourself?

Perhaps?

### Anything else?

can maybe help debug/patch but would need some pointers on where to start to look.

**Comment by nguyenm100** [comment]
if we remove the onotology check statement, rdf4j will return back one row:

query:
```
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix owl: <http://www.w3.org/2002/07/owl#>

select distinct ?modeliri ?modelLabel {
    graph ?modeliri {
        ?modeliri rdfs:label ?modelLabel .
    }
}
```
returns:
```
<http://something/s> | "s-label-something"@en

```

**Comment by pulquero** [comment]
I can confirm this is present in 4.3.3 too (but not in Halyard push strategy :) @nguyenm100)

**Comment by JervenBolleman** [comment]
So the important pieces are that there are two graphs, and that the behaviour is correct when inferencing is off.

**Comment by nguyenm100** [comment]
Hi Jerven, can you elaborate?  Is there a configuration I can use with vanilla rdf4j and workbench the query that gives me the result matching graphdb?  

btw- I created a repo in graphdb w/o inferencing and still received only one result as above.

**Comment by hmottestad** [comment]
One thing worth checking is what happens if all the data is in the unnamed graph.

e.g.
```
<http://something/s> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2002/07/owl#Ontology> .
<http://something/s> <http://www.w3.org/2000/01/rdf-schema#label> "s-label-something"@en .
<http://something/s> <http://www.w3.org/2000/01/rdf-schema#label> "s-label-somethingelse"@en .
```

**Comment by hmottestad** [comment]
Another query we should test:

```
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix owl: <http://www.w3.org/2002/07/owl#>

select distinct ?modeliri ?modelLabel {
    graph ?modeliri {
        ?modeliri a owl:Ontology ;
    }

    graph ?modeliri2 {
        ?modeliri rdfs:label ?modelLabel .
    }

    FILTER(?modeliri = ?modeliri2)
}
```

And maybe also:

```
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix owl: <http://www.w3.org/2002/07/owl#>

select distinct ?modeliri ?modelLabel ?g {
    graph ?g {
        ?modeliri a owl:Ontology ;
        rdfs:label ?modelLabel .
    }
}
```

**Comment by abrokenjester** [comment]
I haven't tested anything but I believe the core of the problem may be that the variable binding the graph name and the subject name is shared. We have a long-standing issue (see https://github.com/eclipse/rdf4j/issues/514) with how context/named graphs are handled in the SPARQL engine, that may be causing the same problem here.

**Comment by abrokenjester** [comment]
Part of the problem, from memory, is that in our current algebra representation, named graph scope is not represented as a separate algebra node, but as an attribute on each statement pattern node. To fix, we should probably refactor the algebra.

**Comment by pulquero** [comment]
Given the bug is not present in my EvalStrategy impl, I'm going to guess there is a bug in StatementPatternQueryEvaluationStep. This could be verified by trying an older version, e.g. v3.7.4.

**Comment by hmottestad** [comment]
I've created a PR to reproduce this issue:

https://github.com/eclipse/rdf4j/pull/4677

**Comment by hmottestad** [comment]
I might have found the issue. There is some code to deduplicate the vars so that we don't insert the same value multiple times into the BindingSet. This code runs before we create the getters for each of the four spoc vars. So in this case we don't create the getter for the context var because it has been set to null during the deduplication. Moving the code for the getters above the code for the deduplication fixes this.

**Comment by hmottestad** [comment]
This has also been fixed in 4.3.4.
